File: 2015/15.py
from Read import *
import re
import math

# Part 1

inp = rdl("15.txt")
#
# inp = """Butterscotch: capacity -1, durability -2, flavor 6, texture 3, calories 8
# Cinnamon: capacity 2, durability 3, flavor -2, texture -1, calories 3""".split('\n')

sweets = dict()

for line in inp:
    ss = line.split(' ')
    sweets[ss[0][:-1]] = {'capacity': int(ss[2][:-1]),
                          'durability': int(ss[4][:-1]),
                          'flavor': int(ss[6][:-1]),
                          'texture': int(ss[8][:-1]),
                          'calories': int(ss[-1])}

max_prod = 0
for i in range(101):
    for j in range(101 - i):
        for k in range(101 - i - j):
            l = 100 - i - j - k

            prod = 1
            for x in ['capacity', 'durability', 'flavor', 'texture']:
                score = sweets['Sprinkles'][x] * i + \
                        sweets['Butterscotch'][x] * j + \
                        sweets['Chocolate'][x] * k + \
                        sweets['Candy'][x] * l

                if score < 0:
                    prod = 0

                prod *= score

            cals = sweets['Sprinkles']['calories'] * i + \
                   sweets['Butterscotch']['calories'] * j + \
                   sweets['Chocolate']['calories'] * k + \
                   sweets['Candy']['calories'] * l

            if prod > max_prod and cals == 500:
                print(prod)
                max_prod = prod

# 10:52 (12th)

# Part 2

# 12:42 (13th)

File: v2/2015/15.R

# Load --------------------------------------------------------------------

rm(list = ls())
source("utilities.R")
inp <- get_input("2015/15", parse = F, user = "blakeebets", cache = T)

# Problem -----------------------------------------------------------------

inp <- strsplit("Butterscotch: capacity -1, durability -2, flavor 6, texture 3, calories 8
Cinnamon: capacity 2, durability 3, flavor -2, texture -1, calories 3", "\n")[[1]]
mat <- as.matrix(as.data.table(expand.grid(a = 0:100, b = 0:100))[a+b == 100])

ss <- strsplit(inp, " ")

stats <- list(as.numeric(gsub(",", "", sapply(ss, function(x) x[3]))),
              as.numeric(gsub(",", "", sapply(ss, function(x) x[5]))),
              as.numeric(gsub(",", "", sapply(ss, function(x) x[7]))),
              as.numeric(gsub(",", "", sapply(ss, function(x) x[9]))),
              as.numeric(gsub(",", "", sapply(ss, function(x) x[11]))))


mat <- as.matrix(as.data.table(expand.grid(a = 0:100, b = 0:100, c = 0:100, d = 0:100))[a+b+c+d == 100])

max(apply(sapply(1:4, function(i) {
    pmax(rowSums(mat * matrix(rep(stats[[i]], nrow(mat)), ncol = ncol(mat), byrow = T)), 0)
}), 1, prod))

# 25:43 (smh), 94th

# Part 2 ------------------------------------------------------------------

# Exactly 500 calories per cookie

dt <- as.data.table(expand.grid(a = 0:100, b = 0:100, c = 0:100, d = 0:100))[a+b+c+d == 100]

dt[, calories := a * stats[[5]][1] + b * stats[[5]][2] + c * stats[[5]][3] + d * stats[[5]][4]]
dt <- dt[calories == 500]

mat <- as.matrix(dt[, -"calories"])

max(apply(sapply(1:4, function(i) {
    pmax(rowSums(mat * matrix(rep(stats[[i]], nrow(mat)), ncol = ncol(mat), byrow = T)), 0)
}), 1, prod))

# 27:12, 81st